### PR TITLE
Refactor APIC register code into an enum

### PR DESCRIPTION
**Pull request by andrisaar**
Unfortunately it looks like supporting x2APIC is not enough. We need to support the older xAPIC as well.

The API is similar enough; the key difference between x2APIC and xAPIC (for us) is that (a) the APIC identifiers are shorter (8 bit vs 32 bit), so we can only support up to 256 CPUs and (b) instead of using MSRs we need to do MMIO to interact with the APIC.

This shuffles some deck chairs around and creates an enum that tries to encapsulate the difference (MSR vs MMIO) while providing an uniform API.

Feel free to bikeshed about the design, though.

### CODE REVIEW

File: stage0/src/apic.rs [hunk 1]
```diff
@@ -19,6 +19,36 @@ use core::arch::x86_64::__cpuid;
...
 use x86_64::{registers::model_specific::Msr, PhysAddr};
 
+pub(crate) enum ApicRegister {
+    X2apic(Msr),
+}
+
+impl ApicRegister {
+    fn read(&self) -> u64 {
+        match self {
+            ApicRegister::X2apic(msr) => unsafe { msr.read() },
```
Comment (conradgrobler) [inline stage0/src/apic.rs:29]: nit: adding some safety comments would be nice. Also, I plan on configuring clippy::undocumented_unsafe_blocks at some point, which would force us to document all unsafe blocks.

**Review by conradgrobler — APPROVED** [review]
I think the design makes sense. Initially I thought about a trait and generics, but since we would need to support all supported APIC types in the same build I don't think that buys us much.

File: stage0/src/apic.rs [hunk 2]
```diff
@@ -82,18 +112,11 @@ impl ApicBase {
 ///
 /// See Section 16.12 (x2APIC_ID) in the AMD64 Architecture Programmer's Manual, Volume 2 for more
 /// details.
-pub struct X2ApicIdRegister;
+pub struct X2ApicIdRegister(ApicRegister);
```
Comment (jul-sh) [inline stage0/src/apic.rs:115]: i personally don't mind this, but afaik in the past we had a convention of not using tuple struct and preferring named properties

**Review by jul-sh — APPROVED** [review]
lgtm

File: stage0/src/apic.rs (hunk 1 above)
Comment (andrisaar) [reply stage0/src/apic.rs:29]: Yep, I agree, I got lazy here.

However, in my furter work I've ended up refactoring a lot of this code anyway, so I'll make sure it has proper comments there and let this slide for now.

Having the Clippy check would be awesome.

File: stage0/src/apic.rs [hunk 3]
```diff
@@ -19,6 +19,36 @@ use core::arch::x86_64::__cpuid;
...
+    fn read(&self) -> u64 {
+        match self {
+            ApicRegister::X2apic(msr) => unsafe { msr.read() },
+        }
+    }
+
+    fn write(&mut self, value: u64) {
+        match self {
+            ApicRegister::X2apic(msr) => unsafe { msr.write(value) },
+        }
```
[the remark this answers is not in the document]
Comment (andrisaar) [reply stage0/src/apic.rs:36]: Yes, we need to support both xAPIC and x2APIC.

**Pull request merged**